### ISSUE TITLE
Upgrading a large number agents results in some getting stuck in "updating"

### ISSUE THREAD

**Issue by ablnk** [elastic/fleet-server#1548]
This defect found during Fleet scale testing. Please note that this defect may be caused by Horde, behavior may differ with real agents. During Fleet Scale sync we agreed to file issues even if it's not clearly a bug.

I reproduced it with 2k and 10k agents.

**Environment:**
ESS cluster was built with oblt-cli.
[Horde](https://github.com/elastic/horde) was used to deploy agents.

**Kibana version:**
8.3.0 SNAPSHOT

**Precondition to reproduce:**
Have 2k or 10k agents in Fleet under the same policy.

**Steps to reproduce:**
1. Click on the select all checkbox. then "Select everything on all pages".
2. Select "Upgrade agents" in actions.
3. Wait for a while and check the result.

**Actual result:**
Agents stuck in "Updating" state.
In logs I only found one error:
```json
{
  "_index": ".ds-elastic-cloud-logs-8-2022.05.28-000004",
  "_id": "LpnsTYEBFt3ZXvKt7Eft",
  "_version": 1,
  "_score": 1,
  "_source": {
    "@timestamp": "2022-06-10T14:03:00.950Z",
    "ecs.version": "1.6.0",
    "service.name": "fleet-server",
    "log": {
      "offset": 1170524,
      "file": {
        "path": "/app/elastic-agent/data/logs/es-containerhost/fleet-server-20220610-4.ndjson"
      }
    },
    "cloud": {
      "availability_zone": "us-west2-b"
    },
    "input": {
      "type": "log"
    },
    "service": {
      "version": "8.3.0-SNAPSHOT",
      "type": "agent",
      "name": "perf-ocuka-custom",
      "id": "f9e9545143faecb03be3530807b8a0b6",
      "node": {
        "name": "instance-0000000000"
      }
    },
    "ecs": {
      "version": "8.0.0"
    },
    "agent": {
      "id": "e2c0815c-7b65-4bb7-82f9-47732f46bf48",
      "name": "f6b2fffd32a5",
      "type": "filebeat",
      "version": "8.3.0",
      "ephemeral_id": "cd40f34b-d445-4e22-9371-5ca7458a9d98"
    },
    "message": "http: TLS handshake error from 10.47.192.54:34866: EOF\n",
    "log.level": "error",
    "event": {
      "dataset": "agent.log"
    },
    "host": {
      "name": "f6b2fffd32a5"
    }
  },
  "fields": {
    "service.name": [
      "perf-ocuka-custom",
      "fleet-server"
    ],
    "service.id": [
      "f9e9545143faecb03be3530807b8a0b6"
    ],
    "service.node.name": [
      "instance-0000000000"
    ],
    "input.type": [
      "log"
    ],
    "log.offset": [
      1170524
    ],
    "message": [
      "http: TLS handshake error from 10.47.192.54:34866: EOF\n"
    ],
    "cloud.availability_zone": [
      "us-west2-b"
    ],
    "service.type": [
      "agent"
    ],
    "agent.type": [
      "filebeat"
    ],
    "@timestamp": [
      "2022-06-10T14:03:00.950Z"
    ],
    "service.version": [
      "8.3.0-SNAPSHOT"
    ],
    "agent.id": [
      "e2c0815c-7b65-4bb7-82f9-47732f46bf48"
    ],
    "ecs.version": [
      "8.0.0",
      "1.6.0"
    ],
    "log.file.path": [
      "/app/elastic-agent/data/logs/es-containerhost/fleet-server-20220610-4.ndjson"
    ],
    "log.level": [
      "error"
    ],
    "agent.ephemeral_id": [
      "cd40f34b-d445-4e22-9371-5ca7458a9d98"
    ],
    "agent.name": [
      "f6b2fffd32a5"
    ],
    "agent.version": [
      "8.3.0"
    ],
    "host.name": [
      "f6b2fffd32a5"
    ],
    "event.dataset": [
      "agent.log"
    ]
  }
}
```

**Expected behavior:**
Selected agents have been upgraded.

**Comment by ablnk** [comment]
FYI @nimarezainia

**Comment by juliaElastic** [comment]
@ablnk for this test, did the `bulk_upgrade` Fleet API call return successfully? Are the `.fleet-actions` and` .fleet-agents` indices updated in kibana as expected?
If so, the issue might be on agent/fleet-server side only.

**Comment by pjbertels** [comment]
Yesterday(June 16th 2022) I tried to update a single Horde drone and it is stuck in 'updating state'. I'm making a change to my cluster that will hopefully get me some more info.

**Comment by ph** [comment]
I am investigating this from https://github.com/elastic/horde/pull/41 too, this might be a regression in horde that I've introduced.

**Comment by juliaElastic** [comment]
I see, that sounds like an issue on agent side, because what Fleet UI does is only update the elastic index, which seems successful (no error in bulk api call).
@nimarezainia is anyone looking at this issue from elastic agent side?

**Comment by ph** [comment]
I am looking from the agent side.

**Comment by pjbertels** [comment]
Based on what I know ... from having done this testing previously, large operations like this get bogged down with resource contention. Some agents complete and some agents get into stuck in ''updating. Now that we have bulk_upgrade maybe the solution is not to allow the workflow Andrei is doing or somehow limit it.

**Comment by pjbertels** [comment]
Debugging session link(pb and ph):
https://app.slack.com/client/T0CUZ52US/C014NMM7BPF/thread/C014NMM7BPF-1655489669.134099

**Comment by nimarezainia** [comment]
I agree with @pjbertels with the rolling upgrades this may not be an issue of concern anymore. I would suggest concentrating our efforts in using the rolling upgrades in the scale test environments.

**Comment by ablnk** [comment]
Update. I tried to upgrade to 8.2.1 instead 8.3. 50% of agents have been updated, 50% stuck in updating state.

**Comment by ph** [comment]
@ablnk Can you give the link to the cluster?

**Comment by ph** [comment]
@ablnk And the which horde installation in GCP so we can check the drone log for that.

**Comment by joshdover** [comment]
> I believe this is due to the agent trying to pull an artifact that doesn't exist, see the thread starting: [elastic/elastic-agent#173 (comment)](https://github.com/elastic/elastic-agent/issues/173#issuecomment-1164263799)

It seems the problem is that the upgrade from the UI on snapshot builds is telling agents to upgrade to `8.3.0` (which isn't released) from production instead of `8.3.0-SNAPSHOT`. I believe this changed from 8.2 and is one unrelated issue that is happening.

> Update. I tried to upgrade to 8.2.1 instead 8.3. 50% of agents have been updated, 50% stuck in updating state.

This seems to be a separate problem, we need to get horde logs to investigate this. I'm also going to run a test tonight and see what results I get.

**Comment by joshdover** [comment]
This can be worked around for now with API calls in Console like:

```
POST kbn:api/fleet/agents/<agentId>/upgrade
{
  "version": "8.3.0-SNAPSHOT",
  "force": true // can be used to force a new upgrade after a failed upgrade
}
```

**Comment by ablnk** [comment]
> @ablnk Can you give the link to the cluster?
https://perf-bscpi-custom.kb.us-west2.gcp.elastic-cloud.com:9243/
horde (http://34.67.34.150/)

**Comment by joshdover** [comment]
I was able to reproduce this and I have some evidence that shows that this is a horde bug, all of the agents that failed have this same `context canceled` error: `Post "https://7a3ba35e101df8cbe368f00531017483.fleet.us-west2.gcp.elastic-cloud.com:443/api/fleet/agents/3327a954-6a42-4673-8278-0d1522874dd5/acks": context canceled`

Will keep investigating.

**Comment by ph** [comment]
Interesting, do you see that the drone is redeployed? Is there more log around that before/after?

**Comment by joshdover** [comment]
@ph has found what we believe to be the root cause of this issue, within horde. When running horde on k8s the autoscaling can cause drones to get shutdown before they've been able to successfully send the "ack" of the upgrade completing and there's no mechanism to retry the ack after they're started again. This causes the agents to stay in the "updating" state.

There is potential that this could affect real Elastic Agent as well, but only in the case that the agent is restarted, which is much less likely (there's no concept of moving an agent from one node to another like there is in our horde emulation). If this were to happen in a production scenario with real agents, I expect it'd only affect a very tiny percentage in the typical case (likely much lower than < 1%). I also believe there is a workaround which would be to force an upgrade to happen again, which is available via the Fleet API. 

We're meeting later today to discuss how to fix this issue for Horde first to unblock testing efforts and also how to fix it with Agent for that edge case scenario.

**Comment by cmacknz** [comment]
>   When running horde on k8s the autoscaling can cause drones to get shutdown before they've been able to successfully send the "ack" of the upgrade completing and there's no mechanism to retry the ack after they're started again.

I think the agent would still be affected when run on Kubernetes for the same reason. The most common reason for this would be the [k8s cluster autoscaler](https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler) that dynamically changes the number of nodes in a cluster. Depending on how it is configured the size of the node pool will fluctuate over the course of a day or week, causing agents in a daemonset to come and go arbitrarily.

**Comment by ph** [comment]
@cmacknz This is a good point, this only affected scheduled and only upgrade which isn't supported in the context of k8s, but as we want to add support for scheduling action to other type of action we will have to consider what happens in the context of k8s. 

Also I am not sure this would apply to k8s, since I'd assume that each new agent moved will be a new agent or ephemeral agent. In the context of horde drone they are permanent and their states is keep between restart.

**Comment by cmacknz** [comment]
Possibly this would affect agents monitoring [StatefulSets](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/):

> Like a [Deployment](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/), a StatefulSet manages Pods that are based on an identical container spec. Unlike a Deployment, a StatefulSet maintains a sticky identity for each of their Pods. These pods are created from the same spec, but are not interchangeable: each has a persistent identifier that it maintains across any rescheduling.
> 
> If you want to use storage volumes to provide persistence for your workload, you can use a StatefulSet as part of the solution. Although individual Pods in a StatefulSet are susceptible to failure, the persistent Pod identifiers make it easier to match existing volumes to the new Pods that replace any that have failed.

Although I'm not sure how DaemonSets interact with StatefulSets. Definitely this is an edge case, but if enough people use agent on k8s they'll hit it.

**Comment by criamico** [comment]
> @criamico Could you confirm whether or not we allow upgrading to `-SNAPSHOT` builds from the UI on snapshot builds of Kibana?

@joshdover  No we left `snapshot` builds out of the list that we show in the UI.

**Comment by pjbertels** [comment]
In the short term to avoid this issue, if we initiate the horde in gke with say a minimum of 4 nodes(in the ehswarm pod). In the case of trying to test 15k drones wouldn't that mitigate the issue? Given the budget is 5000 drones per machine autoscaling would seem not to be involved.

**Comment by joshdover** [comment]
> We should be.
> When we cancel an action, it is removed from the queue, and the (entire) queue is written to the disk in the state file.

Will be addressed in Horde here: https://github.com/elastic/horde/pull/54#discussion_r911286306

We also need to investigate whether or not Agent will remove actions from the queue before they're successfully ack'd. They should only be removed from the queue once they've been completed and ack'd successfully end-to-end.

> In the short term to avoid this issue, if we initiate the horde in gke with say a minimum of 4 nodes(in the ehswarm pod). In the case of trying to test 15k drones wouldn't that mitigate the issue? Given the budget is 5000 drones per machine autoscaling would seem not to be involved if we take this approach.

@pjbertels that will likely work for now. I believe you'd just need to change the `minReplicas` setting on the k8s manifest in horde to 4 (maybe 5 just to be safe?). Let's make sure we remove that and validate the fix in horde once it's been merged.

**Comment by joshdover** [comment]
I wonder if we should add a bit more robustness to this so that losing an ack can't create this scenario at all:
- Keep track of upgrade_started_version: `8.3.0` in the .fleet-agents document
- If the agent checks in and reports that it’s already running this version, then we set upgrade_completed_at as if the ack had happened. This way the agent doesn't keep showing up as "updating" in the UI.

**Comment by joshdover** [comment]
A fix to horde was merged which should help with this case. I still think we should do the robustness improvements in Fleet Server though. @ablnk @pjbertels it'd be good to know if you all ever see this again after that horde change.

**Comment by joshdover** [comment]
I'm going to close this issue for now. We can consider additional robustness (https://github.com/elastic/fleet-server/issues/1548#issuecomment-1178960193) if this presents again in the future. The retry logic in Agent (and now Horde) should handle this already.

**Comment by ablnk** [comment]
I'm reopening the issue since it has been reproduced with 50k agents.

Some observations:
- when reaching 45k upgraded agents, some of them (2k) went offline; at this point, the load on CPU jumped to 61%
![upgrade 50k](https://user-images.githubusercontent.com/34958359/183047954-378adc20-ab31-426b-90a0-98907d4b7ce3.jpg)
- within a few minutes, the number of upgraded (healthy) agents began to grow again
- eventually everything got stuck on the following values: 47611 healthy / 2387 upgrading

Cluster config:
![image](https://user-images.githubusercontent.com/34958359/183048102-dfbdbf84-f0a6-4d7d-b1ee-9ccd3e2e0d66.png)

**Comment by joshdover** [comment]
@pjbertels @ablnk Do we know if this happened with the `upgrade_mode=check_exists` flag on the Horde deployment? I'd like to eliminate any horde-related issues before we go further.

Another thing that would really help us identify where problems are happening is to be looking at the response codes on proxy logs. I am going to see about getting this on the default monitoring dashboards on Cloud.

**Comment by pjbertels** [comment]
We always use upgrade_mode=check_exists flag on the Horde deployment.

**Comment by joshdover** [comment]
Status: @ablnk is going to reproduce this to allow us to look at the logs and metrics in Cloud to do RCA. 

In the future we should keep note of the EC cluster IDs for each component (ES, Kibana, Fleet) for easy lookup after reporting a case (only for test clusters, not for user deployments).

**Comment by joshdover** [comment]
Have we checked to see if we're getting any of these `context canceled` errors we were getting before when drones got rescheduled in horde?

**Comment by joshdover** [comment]
@ablnk would you mind opening a separate issue for this? Though we have the same symptom we were having before at 2k, I think we should treat this as a new investigation so the old information doesn't get confused with the new one.

**Comment by ablnk** [comment]
@joshdover  https://github.com/elastic/fleet-server/issues/1729
